### PR TITLE
created a skeleton function opsim_time_series_images_data

### PR DESCRIPTION
**Pull request by Nikki1510**
Here is my first draft at an opsim skeleton function. Many places still need to be filled out (marked with # To do), but I first want to check if the general idea is okay. I think it makes sense to create a separate function where OpSim Summary is initialised, because that takes a while. It's actually much faster if you have a list of sky coordinates to initialise those together, instead of doing it one coordinate at a time. We can talk about that on Monday. 

To run it, you should have an OpSim database in the data/opsim_database folder, but that one was too large to upload to Github. And you should have a specific branch of OpSim Summary installed (Issue#325/proposalTables). 

Let me know any input you have, I'm happy to change anything, this is only a rough first version :)

### CODE REVIEW

**Comment by Nikki1510** [comment]
It's failing the tests now because I defined some variables that I don't use yet, but that's still on the to-do list. What's the best way to get around this when writing a first code version?

**Comment by ajshajib** [comment]
> It's failing the tests now because I defined some variables that I don't use yet, but that's still on the to-do list. What's the best way to get around this when writing a first code version?

Hi @Nikki1510, the unit tests are passing, which is labeled as "CI / build (3.8)". So, there is no problem with the code itself. However, Pre-commit QA doesn't allow unused variables to maintain a high standard of code formatting. If this PR is still in progress, then presumably, you will use the variables in some future commit within this PR, and these errors will then clear up before the review and merging. However, if the variables are intended to be used in some code that would be provided in a later PR, you can comment out the variable definitions for now. A PR is expected to be one complete set of features, so if variables are meant to be used in a later PR, they should also be defined for the first time within that PR. This helps keep the codebase clean and will make reviewing the future PR easier. I hope this makes sense.

**Comment by Nikki1510** [comment]
@ajshajib Thank you, that makes a lot of sense! I've commented out the variables for now, so that the tests (hopefully) pass, but then this pull request should probably only be approved when I've expanded the function and used all the variables :)

File: slsim/lsst_science_pipeline.py
Comment (sibirrer) [inline slsim/lsst_science_pipeline.py:759]: Can you specify the format a bit more, e.g. what other class needs it and how it should be consistent with it?

File: slsim/lsst_science_pipeline.py
Comment (sibirrer) [inline slsim/lsst_science_pipeline.py:758]: specify that this is in units of pixels

File: slsim/lsst_science_pipeline.py
Comment (sibirrer) [inline slsim/lsst_science_pipeline.py:756]: specify units (deg, arc minuts etc)

File: slsim/lsst_science_pipeline.py
Comment (sibirrer) [inline slsim/lsst_science_pipeline.py:754]: specify format (RA/DEC etc)

File: slsim/lsst_science_pipeline.py
Comment (sibirrer) [inline slsim/lsst_science_pipeline.py:753]: specify a bit more what type of butler instance is needed

File: slsim/lsst_science_pipeline.py
Comment (sibirrer) [inline slsim/lsst_science_pipeline.py:773]: here you can perhaps write it as:
try:
from opsimsummary ....
except:
raise ImportError("Users need to have the right branch of opsimsummary installed (Issue#325/proposalTables)
    # and the opsim database downloaded")
or something of this form

**Review by sibirrer — CHANGES_REQUESTED** [review]
Thank you very much @Nikki1510!
I have some requests for specifying in and output in the documentation. I leave it to @nkhadka21 about the code and if the routines should be split

File: slsim/lsst_science_pipeline.py
Comment (nkhadka21) [inline slsim/lsst_science_pipeline.py:765]: Here, radec is only for 1 observation. During the multiple lens injection, one has to loop this function through list of radec for multiple observations and create a list of astropy table for all observations. Then, one need to pass this list to the multiple_variable_lens_injection() function. So, if opsim is faster with list of radec, I think you can write a function with radec list and pass its output to multiple_variable_lens_injection() directly.

For single radec, you can write another function or give an option within the same function.

**Comment by Nikki1510** [comment]
Hi all, I've updated the function with the PSF map (now a Moffat kernel assuming a power index of 3.1 and fwhm equal to the value from OpSim) and the zero point calculation (taken from OpSimSummary/opsimsummary/simlib.py/add_simlibCols).

**Comment by Nikki1510** [comment]
The output format is different from `dp0_time_series_images_data` because it has the sky brightness instead of `dp0_time_series_cutout`, and it also contains an entry `bandpass` to indicate in which band the observations are taken.

**Comment by nkhadka21** [comment]
Hi @Nikki1510, thank you very much for updating this PR!  It looks good to me. It would be really helpful if you could upload a notebook showing use of your function. You can create a notebook similar to injection_of variable_lenses_in_dp0_time_series_images.ipynb .

**Comment by sibirrer** [comment]
Hi @Nikki1510 just checking in on this PR. I see that in the meantime there were some (minor) conflicts with the main branch. I hope these can be resolved on your end. Let us know if you need input and we are looking forward to having OpSim soon available to simulate light curves of SNe and quasars

**Comment by Nikki1510** [comment]
Hello! I added a notebook ```Opsim_supernovae_plus_extended_source_tutorial.ipynb``` to demonstrate what I have now, but I'm facing a few issues.

- if ```catalog_type=None``` instead of ```"scotch"```, the ```double_sersic``` model is not defined. I get the error: ellipticity or semi-major and semi-minor axis are missing for the first light profile in galaxy_list columns.
- Light curve looks a bit strange for a type Ia SN?
- If the light curve goes down to very low magnitudes (~40), then I get a "lam value too large" error (see screenshot).
- Is it possible to give ```LensPop kwargs_variability={"supernovae_lightcurve", "i"}``` a list of bands instead of just one single band? Because otherwise I cannot compute the brightnesses in the other bands so I cannot simulate multiband images.
- ```lens_class._deflector_dict["mag_i"]}```, and ```lens_class._deflector_dict["z"]``` were not working for me, I replaced them with ```lens_class.deflector_magnitude(band='i')``` and ```lens_class.deflector_redshift```.
- Even when I have a lot of potential lens and source galaxies, the resulting strong lenses are very few (order of 10). A more efficient way to sample would be good.
- The magnitudes of the supernovae are all very faint, so in the images I only see the lens galaxy. I tend to sample very low-redshift lenses and high-redshift sources.

**Comment by nkhadka21** [comment]
Hi @Nikki1510 ,there was an unaccounted factor of  (1+z)**3 in the supernovae number density (conversion from physical to comoving). @cmistick accounted for this, and I believe the issue of a higher number of lensed supernovae at high redshift will be resolved!

**Comment by Nikki1510** [comment]
@nkhadka21 Thank you for your replies! Great to hear LensPop supports multiband, then I had just gotten the format wrong when I tried. And fantastic that the SN number density has been updated, I'll fetch the new changes and run it again :)

**Comment by Nikki1510** [comment]
Update: I fetched the new version of the code (with the corrected (1+z)**3), but I still get only high-redshift supernovae. Also, the amount of strong lensing systems I find is very low. From running the code with an area of 10 square degrees, this is my output:

`34.56 minutes needed to simulate  10 square degrees.`
`Found 12030287 potential lenses and 2233 potential sources.` 
`41.3 minutes needed to generate strong lenses.`
`Number of strong lens systems: 6`

So it takes very long and I'm only left with 6 strong lens systems. The redshifts of the 6 systems are:
`z_l =  [0.59 0.56 0.79 1.04 0.87 0.87]`
`z_s =  [2.1  1.59 1.72 2.33 1.88 1.55]`

I don't have any cuts except for ```kwargs_deflector_cut = {"z_min": 0.01, "z_max": 2.5}```. Maybe this is something we can discuss tomorrow at the SL pipeline meeting, Jacob had an idea how we can allow for multiple sampling methods within the simulation pipeline so then we can also have a faster one.

**Comment by Nikki1510** [comment]
Another issue that we should fix is that currently, it throws an error if a band is outside the wavelength range of a SN. Within my run this happens with the r-band and a SN at redshift 2.1, I get the following error:

`ValueError: bandpass 'lsstr' [5370, .., 7060] outside spectral range [6192.46, .., 61924.6]`

We should think about what we want to happen in this case (append Nans as output? Ignore the band?) so that the code doesn't crash when this happens.

**Comment by sibirrer** [comment]
> Another issue that we should fix is that currently, it throws an error if a band is outside the wavelength range of a SN. Within my run this happens with the r-band and a SN at redshift 2.1, I get the following error:
> 
> `ValueError: bandpass 'lsstr' [5370, .., 7060] outside spectral range [6192.46, .., 61924.6]`
> 
> We should think about what we want to happen in this case (append Nans as output? Ignore the band?) so that the code doesn't crash when this happens.

I add @ahuang314 here to the conversation. I think he could fix it by extending the SNCosmo SN templates.

**Comment by nkhadka21** [comment]
Hi @Nikki1510 , @cmistick has uploaded a notebook showing overview of supernove pop. I think it might be useful while doing supernovae pop. link is here: https://github.com/LSST-strong-lensing/slsim/blob/main/notebooks/supernovae_population_overview.ipynb. It should not take long time to simulate lens population. I think you can specify source_sky_area, deflector_sky_area, and sky_area separately and reduce the time. Carly has done all this in her notebook.

**Comment by ahuang314** [comment]
> Another issue that we should fix is that currently, it throws an error if a band is outside the wavelength range of a SN. Within my run this happens with the r-band and a SN at redshift 2.1, I get the following error:
> 
> `ValueError: bandpass 'lsstr' [5370, .., 7060] outside spectral range [6192.46, .., 61924.6]`
> 
> We should think about what we want to happen in this case (append Nans as output? Ignore the band?) so that the code doesn't crash when this happens.

Hi @Nikki1510 , currently the workaround is to use the extended wavelength supernovae models in the [`data_public`](https://github.com/LSST-strong-lensing/data_public/tree/main/sncosmo_sn_models) repo. In order to use this, download the SALT3.NIR_WAVEEXT directory and supply the path to this directory as the sn_modeldir parameter in your supernova LensPop class. These files are a modified version of the type Ia models which have the lower and upper wavelength ranges extended. If these files are not included, then slsim will use the default models that are built into sncosmo, which can lead to the bandpass error. Unfortunately, these files are too large to be included into slsim directly.

@sibirrer @nkhadka21 As an alternative, I can also work on implementing a feature where a warning is printed and the band is ignored, in case the user is not using the modified SN models or if the bandpass is still outside of the SN wavelength range even with the modified models.

Regarding the strangely shaped lightcurve, this seems to happen for supernovae redshifts above 2. I am not familiar with sncosmo's calculation of the fluxes at higher redshifts, but it seems like the bands are set to flatten out at some value at later times, which can be problematic when the peak at 0 goes lower than this value at high enough redshifts.

**Comment by Nikki1510** [comment]
Hi @cmistick and @ahuang314, thank you so much, that solved the problems! 😄 By adjusting the deflector and source area I can simulate more strong lenses quickly, and using the extended SALT files I can get magnitudes in the bluer filters now as well 👍 I was wondering what you used to extend the SALT template to the bluer wavelengths?

Now I run into another issue: `source.source_dict.colnames` seems to only contain `'ps_mag_g', 'ps_mag_i', 'ps_mag_r'` and not the ones for the z and y bands, even though I initiated LensPop with `kwargs_variability={"supernovae_lightcurve", "g", "r", "i", "z", "y"}`.

**Comment by nkhadka21** [comment]
> Hi @cmistick and @ahuang314, thank you so much, that solved the problems! 😄 By adjusting the deflector and source area I can simulate more strong lenses quickly, and using the extended SALT files I can get magnitudes in the bluer filters now as well 👍 I was wondering what you used to extend the SALT template to the bluer wavelengths?
> 
> Now I run into another issue: `source.source_dict.colnames` seems to only contain `'ps_mag_g', 'ps_mag_i', 'ps_mag_r'` and not the ones for the z and y bands, even though I initiated LensPop with `kwargs_variability={"supernovae_lightcurve", "g", "r", "i", "z", "y"}`.

Hi @Nikki1510 , currently it works only for g, i, and r band in Source class, which is class where light curve extraction happens for individual source. If we want to z and y band, we should add these bands there.

**Comment by ahuang314** [comment]
> I was wondering what you used to extend the SALT template to the bluer wavelengths?

@Nikki1510 Unfortunately, accurate low wavelength models don't yet exist. So, in the past, folks at SNANA just linearly extrapolated the fluxes to zero at 500 angstroms for high z simulations, so that's what I did as well. I have a notebook where I wrote a script to do the extrapolation for me (requires modifying the salt3_template_0.DAT and salt3_template_1.DAT files).

> Now I run into another issue: source.source_dict.colnames seems to only contain 'ps_mag_g', 'ps_mag_i', 'ps_mag_r' and not the ones for the z and y bands, even though I initiated LensPop with kwargs_variability={"supernovae_lightcurve", "g", "r", "i", "z", "y"}.

I just created a PR that adds support for the z and y bands (and also added the warning feature to the Supernova class). Hopefully it can be merged soon

**Comment by Nikki1510** [comment]
> I just created a PR that adds support for the z and y bands (and also added the warning feature to the Supernova class). Hopefully it can be merged soon

@ahuang314 That's fantastic, thank you so much for doing that so quickly! 😄

**Comment by Nikki1510** [comment]
Question: shall I modify `lens_image_series` so that it can take a list of bands (different for each observation), or make a new function `lens_image_series_multiband` that is the same as `lens_image_series` but has input parameter `band` as a list instead of one band?

**Comment by Nikki1510** [comment]
Hi! I pushed a new version of the [notebook](https://github.com/Nikki1510/slsim/blob/opsim/notebooks/Opsim_supernovae_plus_extended_source_tutorial.ipynb), now it works to make time series images of lensed SNe using OpSim 🥳
I want to make another plotting function to display the light curves and the OpSim observations, I'll work on that in the next days. 

I have another question: is it possible to make a cut in SN brightness? I see `kwargs_deflector_cut` and `kwargs_source_cut`, but I guess there's no `kwargs_point_source_cut` yet? Currently I'm getting mostly high-redshift supernovae that are very faint, so it would be good if we could add a brightness cut-off corresponding to the LSST limiting magnitude.

And another one: does the function `sharp_image` also plot the point source? It looks like it does, both in the resulting image and in the code, but in the documentation it says "Creates an unconvolved image of a selected lens. Point source image is not included in this function."

**Comment by nkhadka21** [comment]
> Hi! I pushed a new version of the [notebook](https://github.com/Nikki1510/slsim/blob/opsim/notebooks/Opsim_supernovae_plus_extended_source_tutorial.ipynb), now it works to make time series images of lensed SNe using OpSim 🥳 I want to make another plotting function to display the light curves and the OpSim observations, I'll work on that in the next days.
> 
> I have another question: is it possible to make a cut in SN brightness? I see `kwargs_deflector_cut` and `kwargs_source_cut`, but I guess there's no `kwargs_point_source_cut` yet? Currently I'm getting mostly high-redshift supernovae that are very faint, so it would be good if we could add a brightness cut-off corresponding to the LSST limiting magnitude.
> 
> And another one: does the function `sharp_image` also plot the point source? It looks like it does, both in the resulting image and in the code, but in the documentation it says "Creates an unconvolved image of a selected lens. Point source image is not included in this function."

Hi @Nikki1510 , you can supply magnitude cuts in kwargs_source_cut. This is for both extended source and point source. I have recently made this for both point and extended source. You need to update your branch. However, supernovae catalog currently does not have magnitude. We extract lightcurves only to those supernovae which are in the lensed supernovae population. You might be able to apply LSST cuts in the lensed supernovae. @cmistick is trying to implement supernovae luminosity function in her class which will give us supernovae catalog with redshift and magnitude in i-band. Then, we can be able to apply this cuts in the supernovae catalog.
sharp_image can only simulate image of extended source. We have turned off point source image simulation in this function. For the point source image, we have other functions eg: point_source_image_at_time(). lens_image() and lens_image_series() made use of these functions to simulate full configuration.

**Comment by Nikki1510** [comment]
@nkhadka21 thank you for your replies! 
My output for sharp_image() really does look like it has a point source in it though (see notebook). Or maybe that is the host galaxy?

File: slsim/lsst_science_pipeline.py
Comment (Nikki1510) [reply slsim/lsst_science_pipeline.py]: I think I described the parameters well now

**Comment by Nikki1510** [comment]
Yeeeey all checks passed! 🥳
Would it make sense to merge the pull request now so that everyone can access this? Since I'm going on holiday after tomorrow, and maybe people want to play around with OpSim already. I still want to make another plotting function to nicely display the light curves and LSST observations in each band, but perhaps it's better to make a separate new pull request for that anyway. Let me know what you think, and thank you so much for help answering all my questions along the way! 🙌

**Comment by sibirrer** [comment]
Wohoo!! Thank you very much!
hmm @ajshajib do you know why the test coverage does not appear? 
@Nikki1510 there should be test functions added and passed for these routines (to the extent possible). I am a bit reluctant to merge it to main before that. We can certainly merge it to a branch on the main repository thought

**Comment by ajshajib** [comment]
@sibirrer, I see that the test coverage did update after the most recent commit [9782e9e](https://github.com/LSST-strong-lensing/slsim/pull/125/commits/9782e9e3cfde9b7de44d5bd1c9d1264ee8277192), basically the first message from `CodeCov` above self-updates instead of a new reply after each commit.

**Comment by sibirrer** [comment]
> @sibirrer, I see that the test coverage did update after the most recent commit [9782e9e](https://github.com/LSST-strong-lensing/slsim/pull/125/commits/9782e9e3cfde9b7de44d5bd1c9d1264ee8277192), basically the first message from `CodeCov` above self-updates instead of a new reply after each commit.

ah right, was just too early :)

**Comment by Nikki1510** [comment]
Ah okay, yesterday it seemed to pass the tests but now some checks are not successfull. I'm happy to add test functions, although I don't think I can finish that today. 

The warning seems to be specifally about some small modifications I made to `plot_functions.py`:

```
@codecov
codecov
/ codecov/patch
slsim/Plots/plot_functions.py#L45
Added line #L45 was not covered by tests

@codecov
codecov
/ codecov/patch
slsim/Plots/plot_functions.py#L77
Added line #L77 was not covered by tests
```

Do you know what I should do there to make tests for those lines (or include them in existing tests)? And then more broadly, I assume you'd like to see test functions for the new functions `opsim_time_series_images_data()` and `opsim_variable_lens_injection()`, right? I will need to think a bit about how to do this, if you have ideas for tests then please let me know :)

**Comment by Nikki1510** [comment]
Other things to consider:

- In OpSim there are multiple observations per night, sometimes in the same bands. It could make sense to co-add those (I also did that in my own simulations so I could add a function for that at some point).
- Are the final observed magnitudes already perturbed within the noise level, or are those model magnitudes? I could also add a function to get the limiting flux, perturb the model fluxes, and get the resulting apparent magnitudes and errors.
- Are sky brightness values in units of counts / seconds / arcsec^2 expected to be around  ~0.06? It seems small to me, but that's what I get when doing the conversion from the OpSim sky brightness (will push this new version later today)

**Comment by sibirrer** [comment]
@Nikki1510 for the plotting functions, you can also extend current test functions with an additional example of i.e. providing the argument 'band' as a string (first of your lines) and a similar thing for the second line. It's up to you whether this is a new test function or an extension of a current one (for the first line certainly much easier to do a minor extension).

For the Opsim tests, it's a bit more involved. First you need to check and add (if not done so) the requirement of the OpSim version to SLSim. Then, you should probably have a small 'dummy' OpSim table that you add to the path (can be an input argument) such that the rest of the code can be tested.

File: slsim/lsst_science_pipeline.py
Comment (Nikki1510) [reply slsim/lsst_science_pipeline.py:877]: This is actually solved!! Because Bastien included it in OpSimSummaryV2 😄 Will push a new version now, it's way cleaner 👌

File: slsim/lsst_science_pipeline.py
Comment (Nikki1510) [reply slsim/lsst_science_pipeline.py:810]: Okay good idea. I'm pushing a new version now that has opsim_path as optional input.

File: slsim/lsst_science_pipeline.py
Comment (nkhadka21) [inline slsim/lsst_science_pipeline.py:772]: Hi @Nikki1510, It might be good to mention why you chose these default values (if there is any reason). Also, supernovae lightcurves or quasar lightcurves are in days. So, using MJD time to get magnitude from the interpolated lightcurve might not work. I think we need to change them in days (relative difference).

**Review by nkhadka21 — COMMENTED** [review]
Hi @Nikki1510 , thank you very much for your wonderful effort! I have a one comment about observation time. Please look to that comment and if you have any question, let me know.

**Review by nkhadka21 — CHANGES_REQUESTED** [review]
Hi @Nikki1510, currently, lsst_science_pipeline.py is excluded from testing because it mostly contains functions that cannot be tested locally. However, functions related to opsim can be tested locally. So, I think you could create another file for opsim and write test functions for all those functions. Additionally, you can create a folder named LsstSciencePipeline and place both lsst_science_pipeline.py and your opsim file inside this folder.

**Comment by Nikki1510** [comment]
I tested the test functions on my laptop and they work 👍 
But I don't think they work if the user doesn't have the exact same OpSim database installed... Not sure how to work around that to test the `opsim_time_series_images_data` function. Or could this be enabled as an optional test, only if you want to use OpSim?

**Comment by sibirrer** [comment]
> The second question is about having downloaded the actual OpSim database and stored that in the right folder. That is not on Github but has to be downloaded from another website, and the files are quite large. Similarly, I don't know if we want to require that everyone has such a database downloaded. But if these things are not fulfilled then the test function for the `opsim_time_series_images_data` function will not work.

No, those files should not be required to be downloaded by the user. I would suggest as an input the user gives a path to these files, and if these files are not there (or provided), the code should raise an error with a test that state where the files need to be downloaded from.

**Comment by nkhadka21** [comment]
Hi @Nikki1510 , as Simon mentioned, I think putting OpsimsummaryV2 in test_requirements.txt will solve the first error. For the second error, it seems like the  test function is unable to read your provided path for expo_data_opsim.hdf5 file. Once you provide a correct relative path, then that will solve this error. Thank you.

File: .gitignore
Comment (sibirrer) [inline .gitignore:38]: @Nikki1510 do you have data sets in this folder that are needed for the tests? If so, these are being ignored to be committed. Perhaps this is the reason for the test failures?

File: tests/test_LsstSciencePipeline/test_opsim_pipeline.py
Comment (sibirrer) [inline tests/test_LsstSciencePipeline/test_opsim_pipeline.py:101]: We have used directory paths and then relative to those, i.e.

path = os.path.dirname(__file__)
        module_path, _ = os.path.split(path)
        print(path, module_path)
        blue_one = Table.read(
            os.path.join(path, "TestData/blue_one_modified.fits"), format="fits"
        )

**Comment by sibirrer** [comment]
wohoo, the python problem got resolved. Now it's a weird thing with coolest. @nkhadka21 can you have a look at the CI? Thank you!

**Comment by nkhadka21** [comment]
> wohoo, the python problem got resolved. Now it's a weird thing with coolest. @nkhadka21 can you have a look at the CI? Thank you!

Hi @sibirrer and @Nikki1510, I made an update in slsim coolest interface. Hopefully, this update will solve this error.

**Comment by sibirrer** [comment]
@Nikki1510: @nkhadka21 merged a fix for the coolest part. So if you update your branch, it should hopefully work. We will ignore the pre-commit QA failure for now. Thank you!

**Comment by Nikki1510** [comment]
Hi @sibirrer and @nkhadka21, the CI build was successful now, but the pre-commits are still failing..

**Comment by sibirrer** [comment]
@Nikki1510 we will ignore the recommit-CI tests. We only look at the unit tests and test coverage. I let @nkhadka21 look into whether tests can be improved or not, knowing that we do not need to unit test the OpSim summary part

**Review by nkhadka21 — APPROVED** [review]
Hi @Nikki1510 , thank you very much for your wonderful work! It looks good to me.

**Pull request merged**
